lib/pbench/test/unit/server/test_user_auth.py

```diff
@@ -0,0 +1,365 @@
+import time
+from flask_bcrypt import Bcrypt
+from pbench.server.api.resources.models import UserModel
+
+
+def register_user(
+    client, server_config, email, username, password, firstname, lastname
+):
+    return client.post(
+        f"{server_config.rest_uri}/user",
+        json={
+            "email": email,
+            "password": password,
+            "username": username,
+            "firstName": firstname,
+            "lastName": lastname,
+        },
+    )
+
+
+def login_user(client, server_config, username, password):
+    return client.post(
+        f"{server_config.rest_uri}/session",
+        json={"username": username, "password": password},
+    )
+
+
+class TestUserAuthentication:
+    @staticmethod
+    def test_registration(client, server_config, pytestconfig):
+        client.config["SESSION_FILE_DIR"] = pytestconfig.cache.get("TMP", None)
+        """ Test for user registration """
+        with client:
+            response = register_user(
+                client,
+                server_config,
+                username="user",
+                firstname="firstname",
+                lastname="lastName",
+                email="user@domain.com",
+                password="12345",
+            )
+            data = response.json
+            assert data["status"] == "success"
+            assert data["message"] == "Successfully registered."
+            assert data["auth_token"]
+            assert response.content_type == "application/json"
+            assert response.status_code, 201
+
+    @staticmethod
+    def test_registration_missing_fields(client, server_config):
+        """ Test for user registration missing fields"""
+        with client:
+            response = client.post(
+                f"{server_config.rest_uri}/user",
+                json={
+                    "email": "user@domain.com",
+                    "password": "12345",
+                    "username": "user",
+                },
+            )
+            data = response.json
+            assert data["message"] == "Missing firstName field"
+            assert response.content_type == "application/json"
+            assert response.status_code == 400
+
+    @staticmethod
+    def test_registration_email_validity(client, server_config):
+        """ Test for validating an email field during registration"""
+        with client:
+            response = register_user(
+                client,
+                server_config,
+                username="user",
+                firstname="firstname",
+                lastname="lastName",
+                email="user@domain,com",
```

Comment:
This is one where `@pytest.mark.parametrize` could be useful to test multiple failure patterns easily.

Reply:
yeah thats true